### ISSUE TITLE
Petrostream AI App

### ISSUE THREAD

**Issue by vedthakurchemE** [vedthakurchemE/Semester-1-major-project#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
import base64
from fpdf import FPDF
import io
from io import BytesIO
import streamlit as st
import plotly.express as px

# ─────────────────────────────
# SESSION STATE INIT
# ─────────────────────────────
if "selected_tool" not in st.session_state:
    st.session_state.selected_tool = None

# ─────────────────────────────
# RESET FUNCTIONS
# ─────────────────────────────
def reset_tool():
    st.session_state.clear()

def reset_all():
    for key in list(st.session_state.keys()):
        del st.session_state[key]

if st.button("🔄 Reset All"):
    reset_all()
    st.rerun()

# ─────────────────────────────
# SAMPLE DATA (FOR EXPORT)
# ─────────────────────────────
sample_data = pd.DataFrame({
    "Parameter": ["Pressure", "Temperature", "Flow Rate", "Viscosity"],
    "Value": [250, 98, 1200, 0.89],
    "Unit": ["bar", "°C", "L/h", "Pa.s"]
})

# ─────────────────────────────
# TOOL: SENSOR MONITOR
# ─────────────────────────────
if st.session_state.selected_tool == "Sensor Monitor":
    st.subheader("📡 Real-Time Sensor Monitor")

    fig = px.line(sample_data, x="Parameter", y="Value", title="Sensor Readings")
    st.plotly_chart(fig)

    # Download options
    with st.expander("📤 Download Options"):
        file_format = st.radio("Select File Format", ["CSV", "PDF"])

        if file_format == "CSV":
            csv = sample_data.to_csv(index=False).encode('utf-8')
            st.download_button("⬇️ Download CSV", data=csv, file_name="sensor_data.csv", mime="text/csv")

        elif file_format == "PDF":
            pdf_text = sample_data.to_string()
            pdf_bytes = BytesIO()
            pdf_bytes.write(pdf_text.encode())
            pdf_bytes.seek(0)
            b64_pdf = base64.b64encode(pdf_bytes.read()).decode()
            href = f'<a href="data:application/pdf;base64,{b64_pdf}" download="sensor_data.pdf">⬇️ Download PDF</a>'
            st.markdown(href, unsafe_allow_html=True)

# ─────────────────────────────
# TOOL: MASS TRANSFER COLUMN
# ─────────────────────────────
elif st.session_state.selected_tool == "Mass Transfer Column":
    st.subheader("🌡️ Mass Transfer Column Calculator")
    # Default sample values
    flow_rate = st.number_input("Enter flow rate (mol/s):", value=2.0)
    column_diameter = st.number_input("Enter column diameter (m):", value=0.5)
    height = st.number_input("Enter column height (m):", value=6.0)
    mass_transfer_coeff = st.number_input("Enter mass transfer coefficient (mol/m2.s):", value=0.02)
    area_per_unit_volume = st.number_input("Enter area per unit volume (1/m):", value=100.0)

    # Calculation
    mass_transfer_area = area_per_unit_volume * (3.1416 * (column_diameter ** 2) / 4) * height
    mass_transfer_rate = mass_transfer_coeff * mass_transfer_area

    st.success(f"Estimated Mass Transfer Rate: {mass_transfer_rate:.3f} mol/s")

# ─────────────────────────────
# TOOL: THERMODYNAMICS TOOL
# ─────────────────────────────
elif st.session_state.selected_tool == "Thermodynamics Tool":
    st.subheader("🔥 Thermodynamics Simulator")
    temperature = st.slider("Select Temperature (K):", min_value=250, max_value=1000, value=350)
    pressure = st.slider("Select Pressure (atm):", min_value=1, max_value=100, value=5)
    volume = st.slider("Select Volume (m³):", min_value=1, max_value=10, value=2)
    n_moles = st.number_input("Enter moles of gas (n):", value=1.0)

    # Ideal Gas Law: PV = nRT
    R = 0.0821  # L·atm/(mol·K)
    P_calc = (n_moles * R * temperature) / volume

    st.success(f"Calculated Pressure using Ideal Gas Law: {P_calc:.2f} atm")

# ─────────────────────────────
# RESET BUTTON
# ─────────────────────────────
st.markdown("---")
if st.button("🔁 Reset Tool Selection"):
    reset_tool()
    st.success("Tool reset! Please reselect from dropdown.")
# PDF Generation
pdf = FPDF()
pdf.add_page()
pdf.set_font("Arial", size=12)
pdf.cell(200, 10, txt="PetroStream AI Report", ln=True, align="C")

# Convert to bytes for download
pdf_bytes = pdf.output(dest='S').encode('latin1')
pdf_output = BytesIO(pdf_bytes)

# Download button
st.download_button(
    label="📄 Download Report (PDF)",
    data=pdf_output,
    file_name="report.pdf",
    mime="application/pdf"
)

# ─────────────────────────────────────────────────────────────
# App Config
# ─────────────────────────────────────────────────────────────
st.set_page_config(page_title="PetroStream AI", layout="wide")
st.title("🛢 PetroStream AI: Scalable Digital Twin for Petroleum Optimization")

# Sample Synthetic Petroleum Data
@st.cache_data
def load_data():
    np.random.seed(0)
    df = pd.DataFrame({
        'API_Gravity': np.random.uniform(20, 40, 100),
        'Sulfur_Content': np.random.uniform(0.1, 5.0, 100),
        'Viscosity': np.random.uniform(1, 10, 100),
        'Flowrate': np.random.uniform(100, 1000, 100),
        'Temperature': np.random.uniform(200, 450, 100),
        'Product_Quality': np.random.choice(['High', 'Medium', 'Low'], 100)
    })
    return df

df = load_data()

# ─────────────────────────────────────────────────────────────
# Sidebar Navigation
# ─────────────────────────────────────────────────────────────
option = st.sidebar.radio("🔍 Select Tool", [
    "Thermodynamic Simulator",
    "Mass Transfer Column",
    "Heat Exchanger Model",
    "Petroleum Dataset Analysis",
    "Optimizer",
    "Machine Learning Predictor",
    "Download Reports"
])

# ─────────────────────────────────────────────────────────────
# Reset Button
# ─────────────────────────────────────────────────────────────

# ─────────────────────────────────────────────────────────────
# 1. Thermodynamic System Simulator
# ─────────────────────────────────────────────────────────────
if option == "Thermodynamic Simulator":
    st.subheader("🔥 Thermodynamics Simulator")
    temperature = st.slider("Select Temperature (K):", min_value=250, max_value=1000, value=350)
    pressure = st.slider("Select Pressure (atm):", min_value=1, max_value=100, value=5)
    volume = st.slider("Select Volume (m³):", min_value=1, max_value=10, value=2)
    n_moles = st.number_input("Enter moles of gas (n):", value=1.0)

    # Ideal Gas Law: PV = nRT
    R = 0.0821  # L·atm/(mol·K)
    P_calc = (n_moles * R * temperature) / volume

    st.success(f"Calculated Pressure using Ideal Gas Law: {P_calc:.2f} atm")

# ─────────────────────────────────────────────────────────────
# 2. Mass Transfer Column Modeling
# ─────────────────────────────────────────────────────────────
elif option == "Mass Transfer Column":
    st.subheader("🏗 Distillation Column Simulator")
    xD = st.slider("Distillate Composition", 0.8, 0.99, 0.95)
    xB = st.slider("Bottom Composition", 0.01, 0.2, 0.05)
    q = st.slider("Feed Quality", 0.1, 1.0, 0.5)
    reflux_ratio = st.slider("Reflux Ratio", 1.0, 5.0, 2.0)
    N_stages = int((xD - xB) / (q * reflux_ratio + 0.01) * 5)
    st.info(f"Estimated Number of Stages: {N_stages}")

# ─────────────────────────────────────────────────────────────
# 3. Heat Exchanger Digital Twin
# ─────────────────────────────────────────────────────────────
elif option == "Heat Exchanger Model":
    st.subheader("♨ Heat Exchanger Simulator")
    T_hot_in = st.slider("Hot Fluid Inlet Temp (°C)", 200, 400, 350)
    T_cold_in = st.slider("Cold Fluid Inlet Temp (°C)", 20, 100, 30)
    C_min = 1000
    C_max = 2000
    LMTD = (T_hot_in - T_cold_in) / np.log((T_hot_in - T_cold_in + 1) / 1.1)
    Q = C_min * LMTD
    st.success(f"Heat Duty (Q): {Q:.2f} W")

# ─────────────────────────────────────────────────────────────
# 4. Petroleum Dataset Analyzer
# ─────────────────────────────────────────────────────────────
elif option == "Petroleum Dataset Analysis":
    st.subheader("📊 Petroleum Data Insights")
    st.dataframe(df.head())

    fig, ax = plt.subplots()
    df.groupby("Product_Quality")["API_Gravity"].mean().plot(kind='bar', ax=ax)
    ax.set_ylabel("Avg API Gravity")
    st.pyplot(fig)

# ─────────────────────────────────────────────────────────────
# 5. Optimizer Engine
# ─────────────────────────────────────────────────────────────
elif option == "Optimizer":
    st.subheader("🎯 Optimizer Engine")
    temp = st.slider("Operating Temp (°C)", 100, 500, 300)
    pressure = st.slider("Operating Pressure (bar)", 1, 50, 10)
    flow = st.slider("Flowrate (kg/h)", 100, 1000, 500)
    efficiency = (temp / 500) * (pressure / 50) * (flow / 1000)
    st.success(f"Predicted Yield Efficiency: {efficiency * 100:.2f}%")

# ─────────────────────────────────────────────────────────────
# 6. ML-Based Product Quality Prediction
# ─────────────────────────────────────────────────────────────
elif option == "Machine Learning Predictor":
    st.subheader("🧠 ML Predictor for Product Quality")

    df_ml = df.copy()
    df_ml['Product_Quality'] = df_ml['Product_Quality'].map({'Low': 0, 'Medium': 1, 'High': 2})

    X = df_ml[['API_Gravity', 'Sulfur_Content', 'Viscosity', 'Flowrate', 'Temperature']]
    y = df_ml['Product_Quality']
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

    model = RandomForestClassifier()
    model.fit(X_train, y_train)
    score = model.score(X_test, y_test)

    st.info(f"Model Accuracy: {score*100:.2f}%")

    test_input = [st.slider(i, float(X[i].min()), float(X[i].max()), float(X[i].mean())) for i in X.columns]
    prediction = model.predict([test_input])[0]
    quality_map = {0: 'Low', 1: 'Medium', 2: 'High'}
    st.success(f"Predicted Product Quality: {quality_map[prediction]}")

# ─────────────────────────────────────────────────────────────
# 7. Download CSV / PDF Report
# ─────────────────────────────────────────────────────────────
# ────── Upload CSV ──────
uploaded_file = st.file_uploader("Upload a CSV file", type=["csv"])
if uploaded_file:
    df = pd.read_csv(uploaded_file)
    st.write("📄 Data Preview:", df.head())

    # Select columns
    x_col = st.selectbox("Select X (feature)", df.columns)
    y_col = st.selectbox("Select Y (target)", df.columns)

    if st.button("Run Regression", key="run_regression_btn_3"):
        # Drop rows with NaNs in selected columns
        clean_df = df[[x_col, y_col]].dropna()

        X = clean_df[[x_col]]
        y = clean_df[y_col]

        # Check if X or y are numeric
        if not np.issubdtype(X[x_col].dtype, np.number):
            st.warning("❌ Selected X column must be numeric.")
        elif not np.issubdtype(y.dtype, np.number):
            st.warning("❌ Selected Y column must be numeric.")
        else:
            model = LinearRegression()
            model.fit(X, y)
            y_pred = model.predict(X)

            st.write("### 📈 Regression Results")
            st.write(f"**Intercept:** {model.intercept_:.2f}")
            st.write(f"**Slope:** {model.coef_[0]:.2f}")

            chart_data = pd.DataFrame({x_col: X[x_col], "Actual": y, "Predicted": y_pred})
            st.line_chart(chart_data)

            X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
            model.fit(X_train, y_train)
            score = model.score(X_test, y_test)
            st.success(f"✅ Model R² Score: {score:.4f}")

            st.write("### Regression Results")
            st.write(f"**Intercept:** {model.intercept_:.2f}")
            st.write(f"**Slope:** {model.coef_[0]:.2f}")

            chart_data = pd.DataFrame({x_col: X[x_col], "Actual": y, "Predicted": y_pred})
            st.line_chart(chart_data)

            # Ensure both X and y are numeric
            if not np.issubdtype(X[x_col].dtype, np.number):
                st.error(
                    f"❌ Selected feature column '{x_col}' contains non-numeric data. Please select a numeric column.")
            elif not np.issubdtype(y.dtype, np.number):
                st.error(
                    f"❌ Selected target column '{y_col}' contains non-numeric data. Please select a numeric column.")
            else:
                X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
                model = LinearRegression()
                model.fit(X_train, y_train)
                score = model.score(X_test, y_test)

                st.success(f"✅ Model R² Score: {score:.4f}")

elif option == "Download Reports":
    st.subheader("📥 Download Reports")

    file_choice = st.selectbox("Choose Download Format:", ["CSV", "PDF"])
    if file_choice == "CSV":
        csv = df.to_csv(index=False).encode()
        st.download_button("📁 Download CSV", csv, file_name="petroleum_data.csv", mime='text/csv')
    else:
        try:
            from fpdf import FPDF

            pdf = FPDF()
            pdf.add_page()
            pdf.set_font("Arial", size=12)
            for i, row in df.head(10).iterrows():
                pdf.cell(200, 10, txt=str(row.to_dict()), ln=True)

            pdf_output = io.BytesIO()
            pdf_bytes = pdf.output(dest='S').encode('latin1')  # convert to bytes
            pdf_output = BytesIO(pdf_bytes)
            pdf_output.seek(0)

            b64 = base64.b64encode(pdf_output.read()).decode()
            href = f'<a href="data:application/octet-stream;base64,{b64}" download="petroleum_report.pdf">📄 Download PDF</a>'
            st.markdown(href, unsafe_allow_html=True)
        except ModuleNotFoundError:
            st.warning("Install `fpdf` with `pip install fpdf` for PDF support.")
